File: 2021_Kakao_Internship/red0825/거리두기_확인하기.cpp
#include <string>
#include <vector>

using namespace std;

int d[4][2] = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};

int check(vector<string> s){
    for(int i = 0; i < s.size(); i++){
        for(int j = 0; j < s[i].length(); j++){
            if(s[i][j] == 'P'){
                for(int k = 0; k < 4; k++){
                    int ni = i + d[k][0];
                    int nj = j + d[k][1];
                    if(ni < 0 || ni >= s.size() || nj < 0 || nj >= s[i].length())
                        continue;
                    if(s[ni][nj] == 'P') return 0;
                    if(s[ni][nj] == 'O'){
                        for(int l = 0; l < 4; l++){
                            int nni = ni + d[l][0];
                            int nnj = nj + d[l][1];
                            if(nni < 0 || nni >= s.size() ||
                               nnj < 0 || nnj >= s[i].length() || (nni == i && nnj == j))
                                continue;
                            if(s[nni][nnj] == 'P') return 0;
                        }
                    }
                }
            }
        }
    }
    return 1;
}

vector<int> solution(vector<vector<string>> places) {
    vector<int> answer;
    for(auto x : places){
        answer.push_back(check(x));
    }
    return answer;
}

File: 2021_Kakao_Internship/Hoovely/거리두기 확인하기.py
dx1 = [1, -1, 0, 0]
dy1 = [0, 0, 1, -1]

dx2 = [1, 1, -1, -1]
dy2 = [1, -1, 1, -1]

dx3 = [2, -2, 0, 0]
dy3 = [0, 0, 2, -2]


def func(place):
    for i in range(5):
        for j in range(5):
            if place[i][j] == 'P':
                # 붙어있는 경우
                for k in range(4):
                    x = i + dx1[k]
                    y = j + dy1[k]
                    if 0 <= x < 5 and 0 <= y < 5:
                        if place[x][y] == 'P':
                            return 0
                # 대각선의 경우
                for k in range(4):
                    x = i + dx2[k]
                    y = j + dy2[k]
                    if 0 <= x < 5 and 0 <= y < 5:
                        if place[x][y] == 'P':
                            if place[x][j] != 'X' or place[i][y] != 'X':
                                return 0
                # 2칸씩 떨여져 있는 경우
                for k in range(4):
                    x = i + dx3[k]
                    y = j + dy3[k]
                    if 0 <= x < 5 and 0 <= y < 5:
                        if place[x][y] == 'P':
                            if x == i:
                                if place[i][(y+j)//2] != 'X':
                                    return 0
                            else:
                                if place[(x+i)//2][j] != 'X':
                                    return 0
    return 1


def solution(places):
    answer = []
    for place in places:
        answer.append(func(place))
    return answer
